include/nbl/system/SReadWriteSpinLock.h

```diff
@@ -0,0 +1,98 @@
+#ifndef __NBL_S_READ_WRITE_SPIN_LOCK_H_INCLUDED__
+#define __NBL_S_READ_WRITE_SPIN_LOCK_H_INCLUDED__
+
+#include <atomic>
+#include <thread>
+
+namespace nbl {
+namespace system
+{
+
+namespace impl
+{
+
+    struct SReadWriteSpinLockBase
+    {
+        static inline constexpr uint32_t LockWriteVal = (1u << 31);
+
+        std::atomic_uint32_t m_lock = 0u;
+    };
+
+}
+
+struct SReadWriteSpinLock : protected impl::SReadWriteSpinLockBase
+{
+    void lock_read(std::memory_order order = std::memory_order_seq_cst)
+    {
+        if (m_lock.fetch_add(1u, order) > (LockWriteVal-1u))
+        {
+            while (m_lock.load(order) >= LockWriteVal)
+                std::this_thread::yield();
+        }
+    }
+
+    void unlock_read(std::memory_order order = std::memory_order_seq_cst)
+    {
+        m_lock.fetch_sub(1u, order);
+    }
+
+    void lock_write(std::memory_order order = std::memory_order_seq_cst)
+    {
+        uint32_t expected = 0u;
+        while (!m_lock.compare_exchange_strong(expected, LockWriteVal, order))
+        {
+            expected = 0u;
+            std::this_thread::yield();
```

Comment:
it would be good to spin a few cycles (how many for read and write could be a member var) first before calling `yield` every time

Reply:
please spin a few thousand times before actually yielding.